### PR TITLE
Modify shutdown listeners to support a List of listeners

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
* Modify data prepper core to accept a List of shutdownListeners that will all be run after shutdown.
 
### Issues Resolved
Following from https://github.com/opensearch-project/data-prepper/pull/4185
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by asuresh8 — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java [hunk 1]
```diff
@@ -138,8 +139,8 @@ public Map<String, Pipeline> getTransformationPipelines() {
         return transformationPipelines;
     }
 
-    public void registerShutdownHandler(final DataPrepperShutdownListener shutdownListener) {
-        this.shutdownListener = shutdownListener;
+    public void registerShutdownHandlers(final List<DataPrepperShutdownListener> shutdownListeners) {
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java:142]: Make this take in just a single item, just as it did before. It helps keep the client code cleaner.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/AbstractContextManager.java [hunk 2]
```diff
@@ -59,7 +61,7 @@ private void start() {
         coreApplicationContext.refresh();
 
         dataPrepper = coreApplicationContext.getBean(DataPrepper.class);
-        dataPrepper.registerShutdownHandler(new ContextShutdownListener());
+        dataPrepper.registerShutdownHandlers(List.of(new ContextShutdownListener()));
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/AbstractContextManager.java:64]: You should be able to revert this line based on my comment below.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java [hunk 3]
```diff
@@ -138,8 +139,8 @@ public Map<String, Pipeline> getTransformationPipelines() {
...
+        this.shutdownListeners = shutdownListeners;
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java:143]: This should be implemented as:

```
this.shutdownListeners.add(shutdownListener);
```

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java [hunk 4]
```diff
@@ -45,7 +46,7 @@ public class DataPrepper implements PipelinesProvider {
     @Inject
     @Lazy
     private DataPrepperServer dataPrepperServer;
-    private DataPrepperShutdownListener shutdownListener;
+    private List<DataPrepperShutdownListener> shutdownListeners;
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java:49]:
```
private final List<DataPrepperShutdownListener> shutdownListeners = new LinkedList<>();
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — APPROVED** [review]
Nice, thanks! This is a flexible approach that can be used for other purposes if needed.

**Review by asuresh8 — APPROVED** [review]

**Pull request merged**